Userland/Libraries/LibC/pthread_integration.cpp

```diff
@@ -91,65 +93,95 @@ int __pthread_self()
 
 int pthread_self() __attribute__((weak, alias("__pthread_self")));
 
-int __pthread_mutex_lock(pthread_mutex_t* mutex)
+static constexpr u32 MUTEX_UNLOCKED = 0;
+static constexpr u32 MUTEX_LOCKED_NO_NEED_TO_WAKE = 1;
+static constexpr u32 MUTEX_LOCKED_NEED_TO_WAKE = 2;
+
+int __pthread_mutex_init(pthread_mutex_t* mutex, const pthread_mutexattr_t* attributes)
 {
-    pthread_t this_thread = __pthread_self();
-    for (;;) {
-        u32 expected = 0;
-        if (!AK::atomic_compare_exchange_strong(&mutex->lock, expected, 1u, AK::memory_order_acquire)) {
-            if (mutex->type == __PTHREAD_MUTEX_RECURSIVE && mutex->owner == this_thread) {
-                mutex->level++;
-                return 0;
-            }
-            sched_yield();
-            continue;
-        }
-        mutex->owner = this_thread;
-        mutex->level = 0;
-        return 0;
-    }
+    mutex->lock = 0;
+    mutex->owner = 0;
+    mutex->level = 0;
+    mutex->type = attributes ? attributes->type : __PTHREAD_MUTEX_NORMAL;
+    return 0;
 }
 
-int pthread_mutex_lock(pthread_mutex_t*) __attribute__((weak, alias("__pthread_mutex_lock")));
+int pthread_mutex_init(pthread_mutex_t*, const pthread_mutexattr_t*) __attribute__((weak, alias("__pthread_mutex_init")));
 
-int __pthread_mutex_unlock(pthread_mutex_t* mutex)
+int __pthread_mutex_trylock(pthread_mutex_t* mutex)
 {
-    if (mutex->type == __PTHREAD_MUTEX_RECURSIVE && mutex->level > 0) {
-        mutex->level--;
+    u32 expected = MUTEX_UNLOCKED;
+    bool exchanged = AK::atomic_compare_exchange_strong(&mutex->lock, expected, MUTEX_LOCKED_NO_NEED_TO_WAKE, AK::memory_order_acquire);
+
+    if (exchanged) [[likely]] {
+        mutex->owner = pthread_self();
+        mutex->level = 0;
         return 0;
+    } else if (mutex->type == __PTHREAD_MUTEX_RECURSIVE) {
+        pthread_t owner = AK::atomic_load(&mutex->owner, AK::memory_order_relaxed);
```

Comment:
:thinking: is this necessary?
it's a relaxed load on a value that won't be changed in a way that affects our logic, not sure why it's any better or worse than just reading the owner directly?
(there's probably something I'm missing here...)

Reply:
Sorry, "won't be changed in a way that affects our logic"?

> not sure why it's any better or worse than just reading the owner directly? (there's probably something I'm missing here...)

You're missing that it's UB to access memory non-atomically when nothing else synchronize these accesses. Namely, [cppreference says](https://en.cppreference.com/w/cpp/language/memory_model):

> When an evaluation of an expression writes to a memory location and another evaluation reads or modifies the same memory location, the expressions are said to conflict. A program that has two conflicting evaluations has a _data race_ unless
> * both evaluations execute on the same thread or in the same signal handler, or
> * both conflicting evaluations are atomic operations (see `std::atomic`), or
> * one of the conflicting evaluations happens-before another (see `std::memory_order`)
> If a data race occurs, the behavior of the program is undefined.

So here, it's fine to access the `level` non-atomically, because it's guarded by this same mutex (we only access it while already holding the mutex); but the same isn't true of `owner`.

I actually have left some direct assignments like `mutex->owner = this_thread;` — those need to be fixed, hang on...

Reply:
"our logic" being the loop below, but you're right about UB
That's what I was missing...

Well, at least it found some more UB indirectly lol

Reply:
I did say "some less UB", not "100% UB-free" lol :)